hypershift-operator/controllers/nodepool/inplace.go

```diff
@@ -0,0 +1,402 @@
+	}
+
+	machineName, ok := node.GetAnnotations()[capiv1.MachineAnnotation]
+	if !ok {
+		return nil
+	}
+
+	// Match by namespace when the node has the annotation.
+	machineNamespace, ok := node.GetAnnotations()[capiv1.ClusterNamespaceAnnotation]
+	if !ok {
+		return nil
+	}
+
+	// Match by nodeName and status.nodeRef.name.
+	machine := &capiv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: machineNamespace,
+			Name:      machineName,
+		},
+	}
+	if err := r.Client.Get(context.TODO(), client.ObjectKeyFromObject(machine), machine); err != nil {
+		return nil
+	}
+
+	machineOwner := metav1.GetControllerOf(machine)
+	if machineOwner.Kind != "MachineSet" {
+		return nil
+	}
+
+	machineSet := &capiv1.MachineSet{ObjectMeta: metav1.ObjectMeta{
+		Name:      machineOwner.Name,
+		Namespace: machineNamespace,
+	}}
+	if err := r.Client.Get(context.TODO(), client.ObjectKeyFromObject(machineSet), machineSet); err != nil {
+		return nil
+	}
+
+	nodePoolName := machineSet.GetAnnotations()[nodePoolAnnotation]
+	if nodePoolName == "" {
+		return nil
+	}
+
+	return []reconcile.Request{
+		{NamespacedName: hyperutil.ParseNamespacedName(nodePoolName)},
+	}
+}
+
+func getNodesForMachineSet(ctx context.Context, c client.Reader, hostedClusterClient client.Client, machineSet *capiv1.MachineSet) ([]*corev1.Node, error) {
+	selectorMap, err := metav1.LabelSelectorAsMap(&machineSet.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert MachineSet %q label selector to a map: %v", machineSet.Name, err)
+	}
+
+	// Get all Machines linked to this MachineSet.
+	allMachines := &capiv1.MachineList{}
+	if err = c.List(ctx,
+		allMachines,
+		client.InNamespace(machineSet.Namespace),
+		client.MatchingLabels(selectorMap),
+	); err != nil {
+		return nil, fmt.Errorf("failed to list machines: %v", err)
+	}
+
+	var machineSetOwnedMachines []capiv1.Machine
+	for i, machine := range allMachines.Items {
+		if metav1.GetControllerOf(&machine) != nil && metav1.IsControlledBy(&machine, machineSet) {
+			machineSetOwnedMachines = append(machineSetOwnedMachines, allMachines.Items[i])
+		}
+	}
+
+	var nodes []*corev1.Node
+	for _, machine := range machineSetOwnedMachines {
+		if machine.Status.NodeRef != nil {
+			node := &corev1.Node{}
+			if err := hostedClusterClient.Get(ctx, client.ObjectKey{Name: machine.Status.NodeRef.Name}, node); err != nil {
+				return nil, fmt.Errorf("error getting node: %v", err)
+			}
+			nodes = append(nodes, node)
+		}
+	}
+
+	return nodes, nil
+}
+
+// TODO (alberto): implement.
+func inPlaceUpgradeComplete(nodes []*corev1.Node, targetVersionConfig string) bool {
+	return false
+}
+
+func hostedClusterRESTConfig(ctx context.Context, c client.Reader, name, namespace string) (*restclient.Config, error) {
+	// TODO (alberto): Use a tailored kubeconfig.
+	kubeconfigSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      fmt.Sprintf("%s-admin-kubeconfig", name),
```

Comment:
should you use the reference in the status of the hosted cluster?

Reply:
This is using the kubeconfig that goes through the hosted cluster's external API endpoint. Should it use the one from the service network?

Reply:
updated to use capi kubeconfig for now.